File: edoor/edoor/report/revenue_and_occupancy_summary_report/revenue_and_occupancy_summary_report.js
// Copyright (c) 2023, Tes Pheakdey and contributors
// For license information, please see license.txt
 
frappe.query_reports["Revenue and Occupancy Summary Report"] = {
	"filters": [
		{
			fieldname: "property",
			label: "Property",
			fieldtype: "Link",
			options:"Business Branch",
			default:frappe.defaults.get_user_default("business_branch") ,
			"reqd": 1,
			"on_change": function (query_report) {

				const property = frappe.query_report.get_filter_value("property")
				const business_source_filter =frappe.query_report.get_filter('business_source');
				business_source_filter.df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};
				//room type
				const room_type_filter =frappe.query_report.get_filter('room_type');
				room_type_filter.df.get_query = function() {
					return {
						filters: {
							"property": property
						}
					};
				};

				 
			},
		},
		{
			"fieldname":"start_date",
			"label": __("Start Date"),
			"fieldtype": "Date",
			default: new Date( (new Date()).getFullYear(), ( new Date()).getMonth(), 1),
			"reqd": 1,
			"on_change": function (query_report) {},
		},
		{
			"fieldname":"end_date",
			"label": __("End Date"),
			"fieldtype": "Date",
			default: new Date((new Date()).getFullYear(), (new Date()).getMonth() + 1, 0),
			"on_change": function (query_report) {},
			"reqd": 1
		},

		{
			"fieldname": "room_type",
			"label": __("Room Type"),
			"fieldtype": "Link",
			"options":"Room Type",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "business_source",
			"label": __("Business Source"),
			"fieldtype": "Link",
			"options":"Business Source",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "guest_type",
			"label": __("Guest Type"),
			"fieldtype": "Link",
			"options":"Customer Group",
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "parent_row_group",
			"label": __("Parent Group By"),
			"fieldtype": "Select",
			"options": "\nDate\nMonth\nYear\nReservation Type\nBusiness Source\nBusiness Source Type",
			"on_change": function (query_report) {},
			hide_in_filter:1,
		},
		{
			"fieldname": "row_group",
			"label": __("Row Group By"),
			"fieldtype": "Select",
			"options": "Date\n\Month\nYear\nReservation Type\nBusiness Source\nRoom Type\nRoom\nOutlet\nTable Group\nTable\nPOS Profile\nCustomer\nCustomer Group\nStock Location\nSale Invoice\nWorking Day\nCashier Shift\nSale Type",
			"default":"Date",
			"on_change": function (query_report) {},
			hide_in_filter:1,
		},

		// {
		// 	"fieldname": "column_group",
		// 	"label": __("Column Group By"),
		// 	"fieldtype": "Select",
		// 	"options": "None\nDaily\nWeekly\nMonthly\nQuarterly\nHalf Yearly\nYearly",
		// 	"default":"None"
		// },
		{
			"fieldname": "show_summary",
			"label": __("Show Summary"),
			"fieldtype": "Check",
			"default":1,
			hide_in_filter:1,
			"on_change": function (query_report) {},
		},
		{
			"fieldname": "chart_type",
			"label": __("Chart Type"),
			"fieldtype": "Select",
			"options": "None\nbar\nline\npie",
			"default":"bar",
			hide_in_filter:1,
			"on_change": function (query_report) {},
		},
		 
	],
	onload: function(report) {
		report.page.add_inner_button ("Preview Report", function () {
			frappe.query_report.refresh();
		});
		report.page.add_inner_dropdown ("Preview Report", function () {
			frappe.query_report.refresh();
		});
		
	},
	"formatter": function(value, row, column, data, default_formatter) {
		const origninal_value = value  || 0
		value = default_formatter(value, row, column, data);
		
		
		 
		value = value.toString().replace("style='text-align: right'","style='text-align: " + column.align + "'");	
	 
 
		if (
			(column.fieldtype || "") == "Int" || 
			((column.fieldtype || "") == "Percent") ||
			((column.fieldtype || "") == "Currency" ) 
		) 
		{
			if(origninal_value==0){
				return "-"
			}
		} 


		
		if ((data && data.is_group==1) || (data && data.is_total_row==1)) {
			
			value = $(`<span>${value}</span>`);

			var $value = $(value).css("font-weight", "bold");
			

			value = $value.wrap("<p></p>").parent().html();
		} 
	 
 
		return value;
	},
	
};

 

File: edoor/edoor/report/revenue_and_occupancy_summary_report/revenue_and_occupancy_summary_report.py
# Copyright (c) 2023, Tes Pheakdey and contributors
# For license information, please see license.txt

from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_date
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_month
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_year
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_business_source
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_reservation_type
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_business_source_type
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_guest_type
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_nationality
from edoor.edoor.report.revenue_and_occupancy_summary_report import report_by_room_type
from edoor.api.frontdesk import get_working_day
from frappe import _
from frappe.utils import date_diff,today ,add_months, add_days,getdate,add_to_date
import frappe
from epos_restaurant_2023.utils import get_date_range_by_timespan

def execute(filters=None):
	if not filters.property:
		filters.property = frappe.defaults.get_user_default("business_branch")
	if not filters.property: 
		business_branch = frappe.db.get_list("Business Branch",pluck="name")
		if not filters.property and len(business_branch)>1:
			frappe.throw(_("Please select property"))
		else:
			filters.property = business_branch[0]

	
	if filters.timespan!="Date Range":
		date_range = get_date_range_by_timespan(filters.timespan)
		filters.start_date =date_range["start_date"]
		filters.end_date = date_range["end_date"]
  
  
	if filters.parent_row_group==filters.row_group:
		frappe.throw("Parent row group and row group can not be the same")
		
	report_config = frappe.get_last_doc("Report Configuration", filters={"property":filters.property, "report":"Revenue and Occupancy Summary Report"} )
	
	report = None
	if filters.row_group == "Date":
		report =  report_by_date.get_report(filters, report_config)
	elif filters.row_group == "Month":
		report =  report_by_month.get_report(filters, report_config)
	elif filters.row_group == "Year":
		report =  report_by_year.get_report(filters, report_config)
	elif filters.row_group == "Business Source":
		report =  report_by_business_source.get_report(filters, report_config)	
	elif filters.row_group == "Reservation Type":
		report =  report_by_reservation_type.get_report(filters, report_config)
	elif filters.row_group == "Business Source Type":
		report =  report_by_business_source_type.get_report(filters, report_config)
	elif filters.row_group == "Guest Type":
		report =  report_by_guest_type.get_report(filters, report_config)
	elif filters.row_group == "Nationality":
		report =  report_by_nationality.get_report(filters, report_config)
		
	elif filters.row_group == "Room Type":
		report =  report_by_room_type.get_report(filters, report_config)

	message = _("This is report is for past date transaction")
	return report["columns"], report["data"],message,report["report_chart"], report["report_summary"],True

 
	